Gears/script.cpp

```diff
@@ -1139,60 +1144,181 @@ void functionAShift() {
+        if (currGear < g_vehData.mGearTop) {
+            if (skidding) {
+                // use world speed instead when skiding
+                if (currSpeedWorld > (getGearMaxSpeed(currGear) * 0.9f)) {
+                    shiftTo(g_vehData.mGearCurr + 1, true);
+                    g_gearStates.FakeNeutral = false;
+                    g_gearStates.LastUpshiftTime = currGameTime;
+                }
+            }
+            else {
+                if (g_gearStates.isPowertrainRatioTrustworthy()) {
+                    if (currTotalPower <= (gearPredictStandardPower(currGear + 1) * 0.99f) && currSpeed > (getGearMinSpeed(currGear + 1) * economyCorrection)) {
+                        shiftTo(g_vehData.mGearCurr + 1, true);
+                        g_gearStates.FakeNeutral = false;
+                        g_gearStates.LastUpshiftTime = currGameTime;
+                    }
+                    else if (g_vehData.mRPM > 0.98f && currSpeedWorld > getGearMaxSpeed(currGear)) {
+                        shiftTo(g_vehData.mGearCurr + 1, true);
+                        g_gearStates.FakeNeutral = false;
+                        g_gearStates.LastUpshiftTime = currGameTime;
+                    }
+                }
+                else {
+                    auto currGearMaxSpeed = getGearMaxSpeed(currGear);
+                    auto prevGearMaxSpeed = getGearMaxSpeed(currGear - 1);
+                    if (currSpeed > (((currGearMaxSpeed - prevGearMaxSpeed) * economyCorrection) + (prevGearMaxSpeed * economyCorrection))) {
+                        shiftTo(g_vehData.mGearCurr + 1, true);
+                        g_gearStates.FakeNeutral = false;
+                        g_gearStates.LastUpshiftTime = currGameTime;
+                    }
+                }
+            }
+        }
+        // Shift down
+        if (currGear > 1) {
+            if (skidding) {
+                float currGearMinSpeed = getGearMinSpeed(currGear);
+                float speedGap = (currGearMinSpeed - getGearMinSpeed(currGear - 1)) / 5;
+                if (speedGap < 8) speedGap = 8;
+                if (currSpeedWorld < ((currGearMinSpeed - speedGap) * economyCorrection)) {
+                    shiftTo(currGear - 1, true);
+                    g_gearStates.FakeNeutral = false;
+                }
+            }
+            else {
+                float currGearMinSpeed = getGearMinSpeed(currGear);
+                float speedGap = (currGearMinSpeed - getGearMinSpeed(currGear - 1)) / 5;
+                if (speedGap < 8) speedGap = 8;
+                if (currSpeed < ((currGearMinSpeed - speedGap) * economyCorrection)) {
+                    shiftTo(currGear - 1, true);
+                    g_gearStates.FakeNeutral = false;
+                }
+            }
+        }
+    }
+}
+//For TCU
+float getGearMinSpeed(int gear) {
+    int currGear = g_vehData.mGearCurr;
+    float peakTorqueSpeed = (g_vehData.mDriveMaxFlatVel / g_vehData.mGearRatios[gear]) * 0.75f;
```

Comment:
0.75 might be better suited as a user-configurable (vehicle-specific) number. Especially with some ideas popping up about custom power curves ;)